sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -48,203 +48,239 @@ public class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
-                method => HasArgumentValues(method, "true"), method => IsOverridableDispose(method, containingType));
-        }
+                if (disposeMethodSymbol.IsAbstract ||
+                    disposeMethodSymbol.IsVirtual)
+                {
+                    var modifier = disposeMethod.Modifiers.FirstOrDefault(t => t.IsKind(SyntaxKind.VirtualKeyword) || t.IsKind(SyntaxKind.AbstractKeyword));
 
-        private static bool IsOrImplementsIDisposable(INamedTypeSymbol typeSymbol)
-        {
-            return typeSymbol != null &&
-                (typeSymbol.Is(KnownType.System_IDisposable) || typeSymbol.Implements(KnownType.System_IDisposable));
-        }
+                    AddLocation(modifier.GetLocation(), $"{classSymbol.Name}.Dispose() should not be virtual or abstract.");
+                }
 
-        private static bool IsOrImplementsIDisposable(BaseTypeSyntax baseType, SemanticModel semanticModel)
-        {
-            if (baseType?.Type == null)
+                if (disposeMethodSymbol.ExplicitInterfaceImplementations.Any())
+                {
+                    AddLocation(disposeMethod.Identifier.GetLocation(), $"{classSymbol.Name}.Dispose() should be public.");
+                }
+            }
+
+            private static bool IsVirtualDisposeBool(IMethodSymbol method)
             {
-                return false;
+                return method.Name == DisposeName &&
+                    (method.IsVirtual || method.IsOverride) &&
+                    method.DeclaredAccessibility == Accessibility.Protected &&
+                    method.Parameters.Length == 1 &&
+                    method.Parameters.Any(p => p.Type.Is(KnownType.System_Boolean));
             }
 
-            var typeSymbol = semanticModel.GetSymbolInfo(baseType.Type).Symbol as INamedTypeSymbol;
+            private static bool IsFinalizer(IMethodSymbol method)
+            {
+                return method.MethodKind == MethodKind.Destructor;
+            }
 
-            return typeSymbol != null && typeSymbol.IsInterface() && IsOrImplementsIDisposable(typeSymbol);
-        }
+            private bool IsOrImplementsIDisposable(BaseTypeSyntax baseType)
+            {
+                if (baseType?.Type == null)
+                {
+                    return false;
+                }
 
-        private static bool IsFinalizer(IMethodSymbol method)
-        {
-            return method.MethodKind == MethodKind.Destructor;
-        }
+                var typeSymbol = semanticModel.GetSymbolInfo(baseType.Type).Symbol as INamedTypeSymbol;
 
-        private static bool IsSuppressFinalize(IMethodSymbol method)
-        {
-            return method != null && method.Name == SuppressFinalizeName && method.ContainingType.Is(KnownType.System_GC);
-        }
+                return typeSymbol.IsInterface() &&
+                    typeSymbol.DerivesFrom(KnownType.System_IDisposable);

```

Comment:
`DerivesFrom` for an interface looks very strange. I would use `Is() || Implements()`.